File: weight_tracker.go
# weight_tracker.go
/**
 * ⚖️ Weight Tracker Goals – Fitness Manager (Go Edition)
 * Features: log weight, set goal, progress, chart, stats, BMI, unit switch, JSON storage
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// ─── Types ──────────────────────────────────────────────────────────────────

type Entry struct {
	Date   string  `json:"date"`
	Weight float64 `json:"weight"`
	Notes  string  `json:"notes"`
}

type Data struct {
	GoalWeight   float64   `json:"goalWeight"`
	GoalDeadline string    `json:"goalDeadline"`
	Unit         string    `json:"unit"`
	Height       float64   `json:"height"`
	Entries      []Entry   `json:"entries"`
}

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── Config ──────────────────────────────────────────────────────────────────

const (
	defaultGoal    = 75.0
	defaultUnit    = "kg"
	defaultHeight  = 175.0
)

// ─── Data Manager ──────────────────────────────────────────────────────────

type WeightTracker struct {
	goalWeight   float64
	goalDeadline string
	unit         string
	height       float64
	entries      []Entry
	file         string
	reader       *bufio.Reader
}

func NewWeightTracker() *WeightTracker {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".weight_tracker")
	os.MkdirAll(dir, 0755)
	file := filepath.Join(dir, "data.json")
	w := &WeightTracker{file: file, reader: bufio.NewReader(os.Stdin)}
	w.load()
	return w
}

func (w *WeightTracker) load() {
	if _, err := os.Stat(w.file); os.IsNotExist(err) {
		w.goalWeight = defaultGoal
		w.goalDeadline = defaultDeadline()
		w.unit = defaultUnit
		w.height = defaultHeight
		w.entries = []Entry{}
		return
	}
	raw, err := os.ReadFile(w.file)
	if err != nil {
		w.goalWeight = defaultGoal
		w.goalDeadline = defaultDeadline()
		w.unit = defaultUnit
		w.height = defaultHeight
		w.entries = []Entry{}
		return
	}
	var data Data
	if err := json.Unmarshal(raw, &data); err != nil {
		w.goalWeight = defaultGoal
		w.goalDeadline = defaultDeadline()
		w.unit = defaultUnit
		w.height = defaultHeight
		w.entries = []Entry{}
		return
	}
	w.goalWeight = data.GoalWeight
	if w.goalWeight <= 0 {
		w.goalWeight = defaultGoal
	}
	w.goalDeadline = data.GoalDeadline
	if w.goalDeadline == "" {
		w.goalDeadline = defaultDeadline()
	}
	w.unit = data.Unit
	if w.unit == "" {
		w.unit = defaultUnit
	}
	w.height = data.Height
	if w.height <= 0 {
		w.height = defaultHeight
	}
	w.entries = data.Entries
	if w.entries == nil {
		w.entries = []Entry{}
	}
}

func defaultDeadline() string {
	return time.Now().AddDate(0, 0, 90).Format("2006-01-02")
}

func (w *WeightTracker) save() {
	data := Data{
		GoalWeight:   w.goalWeight,
		GoalDeadline: w.goalDeadline,
		Unit:         w.unit,
		Height:       w.height,
		Entries:      w.entries,
	}
	raw, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(w.file, raw, 0644)
}

func (w *WeightTracker) today() string {
	return time.Now().Format("2006-01-02")
}

func (w *WeightTracker) getTodayEntry() *Entry {
	today := w.today()
	for i := range w.entries {
		if w.entries[i].Date == today {
			return &w.entries[i]
		}
	}
	return nil
}

func (w *WeightTracker) getTodayWeight() *float64 {
	e := w.getTodayEntry()
	if e == nil {
		return nil
	}
	return &e.Weight
}

func (w *WeightTracker) getAllEntriesSorted() []Entry {
	sorted := make([]Entry, len(w.entries))
	copy(sorted, w.entries)
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].Date < sorted[j].Date
	})
	return sorted
}

func (w *WeightTracker) getWeights() []float64 {
	sorted := w.getAllEntriesSorted()
	weights := []float64{}
	for _, e := range sorted {
		weights = append(weights, e.Weight)
	}
	return weights
}

func (w *WeightTracker) progressBar(current, goal float64, width int) string {
	if goal <= 0 {
		return "⚠️  Goal not set"
	}
	weights := w.getWeights()
	if len(weights) == 0 {
		return "No data"
	}
	start := weights[0]
	var ratio float64
	if current >= start {
		ratio = math.Max(0, math.Min(1, (current-goal)/(start-goal)))
	} else {
		ratio = math.Max(0, math.Min(1, (start-current)/(start-goal)))
	}
	filled := int(ratio * float64(width))
	bar := strings.Repeat("█", filled) + strings.Repeat("░", width-filled)
	return fmt.Sprintf("[%s] %.1f%%", bar, ratio*100)
}

func (w *WeightTracker) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := w.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (w *WeightTracker) askFloat(prompt string) float64 {
	for {
		ans := w.ask(prompt)
		if val, err := strconv.ParseFloat(ans, 64); err == nil {
			return val
		}
		fmt.Println(c("❌ Please enter a number.", red))
	}
}

func (w *WeightTracker) askString(prompt string) string {
	return w.ask(prompt)
}

func (w *WeightTracker) askConfirm(prompt string) bool {
	ans := w.ask(prompt + " (yes/no): ")
	return strings.ToLower(ans) == "yes"
}

// ─── Core Actions ──────────────────────────────────────────────────────────

func (w *WeightTracker) addEntry(weight float64, notes string) error {
	if weight <= 0 {
		return fmt.Errorf("weight must be positive")
	}
	today := w.today()
	e := w.getTodayEntry()
	if e != nil {
		e.Weight = weight
		e.Notes = notes
	} else {
		w.entries = append(w.entries, Entry{Date: today, Weight: weight, Notes: notes})
	}
	w.save()
	fmt.Printf(c("✅ Weight logged: %.1f %s\n", green), weight, w.unit)
	return nil
}

func (w *WeightTracker) showToday() {
	weight := w.getTodayWeight()
	if weight == nil {
		fmt.Println(c("No weight logged today.", yellow))
		return
	}
	diff := *weight - w.goalWeight
	status := "above"
	color := red
	if diff < 0 {
		status = "below"
		color = green
	}
	fmt.Println("\n" + c(strings.Repeat("═", 50), dim))
	fmt.Println(c("⚖️ TODAY'S WEIGHT", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), dim))
	fmt.Printf("  Weight: %s\n", c(fmt.Sprintf("%.1f %s", *weight, w.unit), color))
	fmt.Printf("  Goal:   %.1f %s\n", w.goalWeight, w.unit)
	fmt.Printf("  You are %s your goal\n", c(fmt.Sprintf("%.1f %s %s", math.Abs(diff), w.unit, status), color))
	fmt.Printf("  Progress: %s\n", w.progressBar(*weight, w.goalWeight, 20))
	fmt.Println(c(strings.Repeat("═", 50), dim))
}

func (w *WeightTracker) showChart(days int) {
	entries := w.getAllEntriesSorted()
	if len(entries) == 0 {
		fmt.Println(c("No data to chart.", yellow))
		return
	}
	if len(entries) > days {
		entries = entries[len(entries)-days:]
	}
	weights := []float64{}
	dates := []string{}
	for _, e := range entries {
		weights = append(weights, e.Weight)
		dates = append(dates, e.Date[5:10])
	}
	minW, maxW := weights[0], weights[0]
	for _, w := range weights {
		if w < minW {
			minW = w
		}
		if w > maxW {
			maxW = w
		}
	}
	rangeW := maxW - minW
	if rangeW == 0 {
		fmt.Println(c("Weight is constant. No variation to chart.", yellow))
		return
	}
	height := 10
	chartWidth := 40
	norm := make([]int, len(weights))
	for i, w := range weights {
		norm[i] = int((w - minW) / rangeW * float64(height-1))
	}
	lines := []string{}
	for row := height - 1; row >= 0; row-- {
		line := ""
		for i, val := range norm {
			if val >= row {
				if i > 0 && norm[i-1] >= row {
					line += "─"
				} else {
					line += "┌"
				}
			} else {
				line += " "
			}
		}
		lines = append(lines, line)
	}
	step := len(dates) / 8
	if step < 1 {
		step = 1
	}
	xAxis := " "
	lastPos := 0
	for i := 0; i < len(dates); i += step {
		if i >= len(dates) {
			break
		}
		label := dates[i]
		pos := i
		if pos > lastPos {
			xAxis += strings.Repeat(" ", pos-lastPos)
		}
		xAxis += label
		lastPos = pos
	}
	if lastPos < len(dates)-1 {
		xAxis += strings.Repeat(" ", len(dates)-1-lastPos)
	}
	fmt.Printf("\n%s\n", c(fmt.Sprintf("📈 Weight Chart (last %d days)", len(entries)), bright+cyan))
	fmt.Println(strings.Join(lines, "\n"))
	fmt.Println(xAxis)
	fmt.Printf("Min: %.1f %s  Max: %.1f %s\n", minW, w.unit, maxW, w.unit)
}

func (w *WeightTracker) showStats() {
	if len(w.entries) == 0 {
		fmt.Println(c("📭 No data yet. Start tracking!", yellow))
		return
	}
	weights := w.getWeights()
	if len(weights) == 0 {
		fmt.Println(c("No weight data available.", yellow))
		return
	}
	total := len(weights)
	sum := 0.0
	for _, v := range weights {
		sum += v
	}
	avg := sum / float64(total)
	minW, maxW := weights[0], weights[0]
	for _, v := range weights {
		if v < minW {
			minW = v
		}
		if v > maxW {
			maxW = v
		}
	}
	last := weights[len(weights)-1]
	first := weights[0]
	change := last - first
	trend := "down"
	if change > 0 {
		trend = "up"
	}
	heightM := w.height / 100.0
	var bmi float64
	if heightM > 0 {
		bmi = last / (heightM * heightM)
	}
	bmiCat := bmiCategory(bmi)
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Total Entries:  %d\n", total)
	fmt.Printf("  Current Weight: %s\n", c(fmt.Sprintf("%.1f %s", last, w.unit), green))
	fmt.Printf("  Goal Weight:    %.1f %s\n", w.goalWeight, w.unit)
	fmt.Printf("  Average:        %.1f %s\n", avg, w.unit)
	fmt.Printf("  Minimum:        %.1f %s\n", minW, w.unit)
	fmt.Printf("  Maximum:        %.1f %s\n", maxW, w.unit)
	sign := ""
	if change > 0 {
		sign = "+"
	}
	fmt.Printf("  Change:         %s%.1f %s (%s)\n", sign, change, w.unit, trend)
	fmt.Printf("  BMI:            %.1f (%s)\n", bmi, bmiCat)
}

func bmiCategory(bmi float64) string {
	if bmi < 18.5 {
		return "Underweight"
	}
	if bmi < 25 {
		return "Normal"
	}
	if bmi < 30 {
		return "Overweight"
	}
	return "Obese"
}

func (w *WeightTracker) setGoal(weight float64, deadline string) error {
	if weight <= 0 {
		return fmt.Errorf("goal weight must be positive")
	}
	w.goalWeight = weight
	if deadline != "" {
		if _, err := time.Parse("2006-01-02", deadline); err != nil {
			return fmt.Errorf("invalid date format, use YYYY-MM-DD")
		}
		w.goalDeadline = deadline
	} else {
		w.goalDeadline = defaultDeadline()
	}
	w.save()
	fmt.Printf(c("✅ Goal set to %.1f %s by %s\n", green), weight, w.unit, w.goalDeadline)
	return nil
}

func (w *WeightTracker) setHeight(height float64) error {
	if height <= 0 {
		return fmt.Errorf("height must be positive")
	}
	w.height = height
	w.save()
	fmt.Printf(c("✅ Height set to %.1f cm\n", green), height)
	return nil
}

func (w *WeightTracker) setUnit(unit string) error {
	if unit != "kg" && unit != "lbs" {
		return fmt.Errorf("unit must be 'kg' or 'lbs'")
	}
	w.unit = unit
	w.save()
	fmt.Printf(c("✅ Unit set to %s\n", green), unit)
	return nil
}

func (w *WeightTracker) clearData() {
	if !w.askConfirm("⚠️  Delete ALL data? This cannot be undone!") {
		return
	}
	w.entries = []Entry{}
	w.goalWeight = defaultGoal
	w.goalDeadline = defaultDeadline()
	w.save()
	fmt.Println(c("🗑️  All data cleared.", yellow))
}

// ─── Menu ──────────────────────────────────────────────────────────────────

func (w *WeightTracker) showMenu() {
	todayWeight := w.getTodayWeight()
	var weightStr string
	if todayWeight != nil {
		weightStr = fmt.Sprintf("%.1f", *todayWeight)
	} else {
		weightStr = "—"
	}
	progress := w.progressBar(func() float64 {
		if todayWeight != nil {
			return *todayWeight
		}
		return 0
	}(), w.goalWeight, 20)
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("⚖️ WEIGHT TRACKER", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Today: %s %s / %.1f %s\n", weightStr, w.unit, w.goalWeight, w.unit)
	fmt.Printf("  Goal deadline: %s\n", w.goalDeadline)
	fmt.Printf("  Progress: %s\n", progress)
	fmt.Println(c(strings.Repeat("─", 50), dim))
	fmt.Println("  1. ⚖️ Log weight today")
	fmt.Println("  2. 📊 Today's progress")
	fmt.Println("  3. 📈 Show weight chart")
	fmt.Println("  4. 📊 Statistics")
	fmt.Println("  5. 🎯 Set goal weight & deadline")
	fmt.Println("  6. 📏 Set height")
	fmt.Println("  7. 🔄 Set unit (kg/lbs)")
	fmt.Println("  8. 🗑️  Clear all data")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (w *WeightTracker) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n⚖️ Weight Tracker – Goal‑Based Fitness Manager", bright+cyan))
	fmt.Println(c("Track your weight, reach your goals!", dim))

	for {
		w.showMenu()
		choice := w.ask("Your choice: ")
		switch choice {
		case "1":
			weight := w.askFloat(fmt.Sprintf("Weight (%s): ", w.unit))
			notes := w.askString("Notes (optional): ")
			w.addEntry(weight, notes)
		case "2":
			w.showToday()
		case "3":
			w.showChart(30)
		case "4":
			w.showStats()
		case "5":
			weight := w.askFloat(fmt.Sprintf("Goal weight (%s): ", w.unit))
			deadline := w.askString("Deadline (YYYY-MM-DD) [leave empty for default]: ")
			if err := w.setGoal(weight, deadline); err != nil {
				fmt.Printf("%s\n", c("❌ "+err.Error(), red))
			}
		case "6":
			height := w.askFloat("Height (cm): ")
			if err := w.setHeight(height); err != nil {
				fmt.Printf("%s\n", c("❌ "+err.Error(), red))
			}
		case "7":
			unit := w.askString("Unit (kg/lbs): ")
			if err := w.setUnit(strings.ToLower(unit)); err != nil {
				fmt.Printf("%s\n", c("❌ "+err.Error(), red))
			}
		case "8":
			w.clearData()
		case "0":
			fmt.Println(c("👋 Stay fit! Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			w.reader.ReadString('\n')
		}
	}
}

func main() {
	app := NewWeightTracker()
	app.run()
}

File: weight_tracker.rs
# weight_tracker.rs
/**
 * ⚖️ Weight Tracker Goals – Fitness Manager (Rust Edition)
 * Features: log weight, set goal, progress, chart, stats, BMI, unit switch, JSON storage
 * Dependencies: serde, serde_json, chrono, colored
 */

use chrono::{Local, Duration};
use colored::*;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::PathBuf;

// ─── Types ──────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Entry {
    date: String,
    weight: f64,
    notes: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct Data {
    goal_weight: f64,
    goal_deadline: String,
    unit: String,
    height: f64,
    entries: Vec<Entry>,
}

// ─── Colors ──────────────────────────────────────────────────────────────────

fn c(text: &str, color: &str) -> String {
    match color {
        "green" => text.green().to_string(),
        "red" => text.red().to_string(),
        "yellow" => text.yellow().to_string(),
        "cyan" => text.cyan().to_string(),
        "bright" => text.bright().to_string(),
        "dim" => text.dimmed().to_string(),
        _ => text.to_string(),
    }
}

// ─── Config ──────────────────────────────────────────────────────────────────

const DEFAULT_GOAL: f64 = 75.0;
const DEFAULT_UNIT: &str = "kg";
const DEFAULT_HEIGHT: f64 = 175.0;

// ─── Data Manager ──────────────────────────────────────────────────────────

struct WeightTracker {
    goal_weight: f64,
    goal_deadline: String,
    unit: String,
    height: f64,
    entries: Vec<Entry>,
    file_path: PathBuf,
}

impl WeightTracker {
    fn new() -> Self {
        let home = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE")).unwrap_or_else(|_| ".".to_string());
        let dir = PathBuf::from(home).join(".weight_tracker");
        fs::create_dir_all(&dir).unwrap();
        let file_path = dir.join("data.json");
        let mut wt = WeightTracker {
            goal_weight: DEFAULT_GOAL,
            goal_deadline: default_deadline(),
            unit: DEFAULT_UNIT.to_string(),
            height: DEFAULT_HEIGHT,
            entries: Vec::new(),
            file_path,
        };
        wt.load();
        wt
    }

    fn load(&mut self) {
        if let Ok(raw) = fs::read_to_string(&self.file_path) {
            if let Ok(data) = serde_json::from_str::<Data>(&raw) {
                self.goal_weight = if data.goal_weight > 0.0 { data.goal_weight } else { DEFAULT_GOAL };
                self.goal_deadline = if data.goal_deadline.is_empty() { default_deadline() } else { data.goal_deadline };
                self.unit = if data.unit.is_empty() { DEFAULT_UNIT.to_string() } else { data.unit };
                self.height = if data.height > 0.0 { data.height } else { DEFAULT_HEIGHT };
                self.entries = data.entries;
                return;
            }
        }
        self.goal_weight = DEFAULT_GOAL;
        self.goal_deadline = default_deadline();
        self.unit = DEFAULT_UNIT.to_string();
        self.height = DEFAULT_HEIGHT;
        self.entries = Vec::new();
    }

    fn save(&self) {
        let data = Data {
            goal_weight: self.goal_weight,
            goal_deadline: self.goal_deadline.clone(),
            unit: self.unit.clone(),
            height: self.height,
            entries: self.entries.clone(),
        };
        let raw = serde_json::to_string_pretty(&data).unwrap();
        let _ = fs::write(&self.file_path, raw);
    }

    fn today(&self) -> String {
        Local::now().format("%Y-%m-%d").to_string()
    }

    fn get_today_entry(&mut self) -> Option<&mut Entry> {
        let today = self.today();
        self.entries.iter_mut().find(|e| e.date == today)
    }

    fn get_today_weight(&self) -> Option<f64> {
        let today = self.today();
        self.entries.iter().find(|e| e.date == today).map(|e| e.weight)
    }

    fn get_all_entries_sorted(&self) -> Vec<Entry> {
        let mut sorted = self.entries.clone();
        sorted.sort_by(|a, b| a.date.cmp(&b.date));
        sorted
    }

    fn get_weights(&self) -> Vec<f64> {
        self.get_all_entries_sorted().iter().map(|e| e.weight).collect()
    }

    fn progress_bar(&self, current: f64, goal: f64, width: usize) -> String {
        if goal <= 0.0 {
            return "⚠️  Goal not set".to_string();
        }
        let weights = self.get_weights();
        if weights.is_empty() {
            return "No data".to_string();
        }
        let start = weights[0];
        let ratio = if current >= start {
            ((current - goal) / (start - goal)).max(0.0).min(1.0)
        } else {
            ((start - current) / (start - goal)).max(0.0).min(1.0)
        };
        let filled = (ratio * width as f64) as usize;
        let bar = "█".repeat(filled) + &"░".repeat(width - filled);
        format!("[{}] {:.1}%", bar, ratio * 100.0)
    }

    fn ask(&self, prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        let mut line = String::new();
        io::stdin().read_line(&mut line).unwrap();
        line.trim().to_string()
    }

    fn ask_float(&self, prompt: &str) -> f64 {
        loop {
            let ans = self.ask(prompt);
            if let Ok(val) = ans.parse::<f64>() {
                return val;
            }
            println!("{}", c("❌ Please enter a number.", "red"));
        }
    }

    fn ask_confirm(&self, prompt: &str) -> bool {
        let ans = self.ask(&format!("{} (yes/no): ", prompt));
        ans.to_lowercase() == "yes"
    }

    // ─── Core Actions ──────────────────────────────────────────────────────

    fn add_entry(&mut self, weight: f64, notes: String) -> Result<(), Box<dyn std::error::Error>> {
        if weight <= 0.0 {
            return Err("Weight must be positive!".into());
        }
        let today = self.today();
        if let Some(entry) = self.get_today_entry() {
            entry.weight = weight;
            entry.notes = notes;
        } else {
            self.entries.push(Entry { date: today, weight, notes });
        }
        self.save();
        println!("{}", c(&format!("✅ Weight logged: {:.1} {}", weight, self.unit), "green"));
        Ok(())
    }

    fn show_today(&self) {
        if let Some(weight) = self.get_today_weight() {
            let diff = weight - self.goal_weight;
            let status = if diff > 0.0 { "above" } else { "below" };
            let color = if diff > 0.0 { "red" } else { "green" };
            println!("\n{}", "═".repeat(50).dimmed());
            println!("{}", c("⚖️ TODAY'S WEIGHT", "bright cyan"));
            println!("{}", "═".repeat(50).dimmed());
            println!("  Weight: {}", c(&format!("{:.1} {}", weight, self.unit), color));
            println!("  Goal:   {:.1} {}", self.goal_weight, self.unit);
            println!("  You are {} your goal", c(&format!("{:.1} {} {}", diff.abs(), self.unit, status), color));
            println!("  Progress: {}", self.progress_bar(weight, self.goal_weight, 20));
            println!("{}", "═".repeat(50).dimmed());
        } else {
            println!("{}", c("No weight logged today.", "yellow"));
        }
    }

    fn show_chart(&self, days: usize) {
        let mut entries = self.get_all_entries_sorted();
        if entries.is_empty() {
            println!("{}", c("No data to chart.", "yellow"));
            return;
        }
        if entries.len() > days {
            entries = entries.into_iter().skip(entries.len() - days).collect();
        }
        let weights: Vec<f64> = entries.iter().map(|e| e.weight).collect();
        let dates: Vec<String> = entries.iter().map(|e| e.date[5..10].to_string()).collect();
        let min_w = weights.iter().cloned().fold(f64::INFINITY, f64::min);
        let max_w = weights.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        let range = max_w - min_w;
        if range == 0.0 {
            println!("{}", c("Weight is constant. No variation to chart.", "yellow"));
            return;
        }
        let height = 10;
        let chart_width = 40;
        let norm: Vec<usize> = weights.iter().map(|w| ((w - min_w) / range * (height - 1) as f64) as usize).collect();
        let mut lines = Vec::new();
        for row in (0..height).rev() {
            let mut line = String::new();
            for (i, &val) in norm.iter().enumerate() {
                if val >= row {
                    if i > 0 && norm[i-1] >= row {
                        line.push('─');
                    } else {
                        line.push('┌');
                    }
                } else {
                    line.push(' ');
                }
            }
            lines.push(line);
        }
        let step = (dates.len() / 8).max(1);
        let mut x_axis = " ".to_string();
        let mut last_pos = 0;
        for (i, label) in dates.iter().enumerate().step_by(step) {
            let pos = i;
            if pos > last_pos {
                x_axis.push_str(&" ".repeat(pos - last_pos));
            }
            x_axis.push_str(label);
            last_pos = pos;
        }
        if last_pos < dates.len() - 1 {
            x_axis.push_str(&" ".repeat(dates.len() - 1 - last_pos));
        }
        println!("\n{}", c(&format!("📈 Weight Chart (last {} days)", entries.len()), "bright cyan"));
        println!("{}", lines.join("\n"));
        println!("{}", x_axis);
        println!("Min: {:.1} {}  Max: {:.1} {}", min_w, self.unit, max_w, self.unit);
    }

    fn show_stats(&self) {
        if self.entries.is_empty() {
            println!("{}", c("📭 No data yet. Start tracking!", "yellow"));
            return;
        }
        let weights = self.get_weights();
        if weights.is_empty() {
            println!("{}", c("No weight data available.", "yellow"));
            return;
        }
        let total = weights.len();
        let sum: f64 = weights.iter().sum();
        let avg = sum / total as f64;
        let min_w = weights.iter().cloned().fold(f64::INFINITY, f64::min);
        let max_w = weights.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        let last = weights[weights.len()-1];
        let first = weights[0];
        let change = last - first;
        let trend = if change > 0.0 { "up" } else { "down" };
        let height_m = self.height / 100.0;
        let bmi = if height_m > 0.0 { last / (height_m * height_m) } else { 0.0 };
        let bmi_cat = bmi_category(bmi);
        println!("\n📊 STATISTICS");
        println!("{}", "─".repeat(30).dimmed());
        println!("  Total Entries:  {}", total);
        println!("  Current Weight: {}", c(&format!("{:.1} {}", last, self.unit), "green"));
        println!("  Goal Weight:    {:.1} {}", self.goal_weight, self.unit);
        println!("  Average:        {:.1} {}", avg, self.unit);
        println!("  Minimum:        {:.1} {}", min_w, self.unit);
        println!("  Maximum:        {:.1} {}", max_w, self.unit);
        let sign = if change > 0.0 { "+" } else { "" };
        println!("  Change:         {}{:.1} {} ({})", sign, change, self.unit, trend);
        println!("  BMI:            {:.1} ({})", bmi, bmi_cat);
    }

    fn set_goal(&mut self, weight: f64, deadline: Option<String>) -> Result<(), Box<dyn std::error::Error>> {
        if weight <= 0.0 {
            return Err("Goal weight must be positive!".into());
        }
        self.goal_weight = weight;
        if let Some(d) = deadline {
            if chrono::NaiveDate::parse_from_str(&d, "%Y-%m-%d").is_err() {
                return Err("Invalid date format. Use YYYY-MM-DD".into());
            }
            self.goal_deadline = d;
        } else {
            self.goal_deadline = default_deadline();
        }
        self.save();
        println!("{}", c(&format!("✅ Goal set to {:.1} {} by {}", weight, self.unit, self.goal_deadline), "green"));
        Ok(())
    }

    fn set_height(&mut self, height: f64) -> Result<(), Box<dyn std::error::Error>> {
        if height <= 0.0 {
            return Err("Height must be positive!".into());
        }
        self.height = height;
        self.save();
        println!("{}", c(&format!("✅ Height set to {:.1} cm", height), "green"));
        Ok(())
    }

    fn set_unit(&mut self, unit: String) -> Result<(), Box<dyn std::error::Error>> {
        if unit != "kg" && unit != "lbs" {
            return Err("Unit must be 'kg' or 'lbs'".into());
        }
        self.unit = unit;
        self.save();
        println!("{}", c(&format!("✅ Unit set to {}", self.unit), "green"));
        Ok(())
    }

    fn clear_data(&mut self) {
        if !self.ask_confirm("⚠️  Delete ALL data? This cannot be undone!") {
            return;
        }
        self.entries = Vec::new();
        self.goal_weight = DEFAULT_GOAL;
        self.goal_deadline = default_deadline();
        self.save();
        println!("{}", c("🗑️  All data cleared.", "yellow"));
    }

    // ─── Menu ──────────────────────────────────────────────────────────────

    fn show_menu(&self) {
        let today_weight = self.get_today_weight();
        let weight_str = if let Some(w) = today_weight { format!("{:.1}", w) } else { "—".to_string() };
        let progress = self.progress_bar(today_weight.unwrap_or(0.0), self.goal_weight, 20);
        println!("\n{}", "═".repeat(50).cyan());
        println!("{}", c("⚖️ WEIGHT TRACKER", "bright cyan"));
        println!("{}", "═".repeat(50).cyan());
        println!("  Today: {} {} / {:.1} {}", weight_str, self.unit, self.goal_weight, self.unit);
        println!("  Goal deadline: {}", self.goal_deadline);
        println!("  Progress: {}", progress);
        println!("{}", "─".repeat(50).dimmed());
        println!("  1. ⚖️ Log weight today");
        println!("  2. 📊 Today's progress");
        println!("  3. 📈 Show weight chart");
        println!("  4. 📊 Statistics");
        println!("  5. 🎯 Set goal weight & deadline");
        println!("  6. 📏 Set height");
        println!("  7. 🔄 Set unit (kg/lbs)");
        println!("  8. 🗑️  Clear all data");
        println!("  0. 🚪 Exit");
        println!("{}", "═".repeat(50).cyan());
    }

    fn run(&mut self) {
        println!("{}", "\n⚖️ Weight Tracker – Goal‑Based Fitness Manager".bright().cyan());
        println!("{}", "Track your weight, reach your goals!".dimmed());

        loop {
            self.show_menu();
            let choice = self.ask("Your choice: ");
            match choice.as_str() {
                "1" => {
                    let weight = self.ask_float(&format!("Weight ({}): ", self.unit));
                    let notes = self.ask("Notes (optional): ");
                    if let Err(e) = self.add_entry(weight, notes) {
                        println!("{}", c(&format!("❌ {}", e), "red"));
                    }
                }
                "2" => self.show_today(),
                "3" => self.show_chart(30),
                "4" => self.show_stats(),
                "5" => {
                    let weight = self.ask_float(&format!("Goal weight ({}): ", self.unit));
                    let deadline = self.ask("Deadline (YYYY-MM-DD) [leave empty for default]: ");
                    if let Err(e) = self.set_goal(weight, if deadline.is_empty() { None } else { Some(deadline) }) {
                        println!("{}", c(&format!("❌ {}", e), "red"));
                    }
                }
                "6" => {
                    let height = self.ask_float("Height (cm): ");
                    if let Err(e) = self.set_height(height) {
                        println!("{}", c(&format!("❌ {}", e), "red"));
                    }
                }
                "7" => {
                    let unit = self.ask("Unit (kg/lbs): ");
                    if let Err(e) = self.set_unit(unit) {
                        println!("{}", c(&format!("❌ {}", e), "red"));
                    }
                }
                "8" => self.clear_data(),
                "0" => {
                    println!("{}", c("👋 Stay fit! Goodbye!", "cyan"));
                    return;
                }
                _ => println!("{}", c("❌ Invalid choice.", "red")),
            }
            if choice != "0" {
                print!("\nPress Enter to continue...");
                io::stdout().flush().unwrap();
                let mut _dummy = String::new();
                io::stdin().read_line(&mut _dummy).unwrap();
            }
        }
    }
}

fn default_deadline() -> String {
    (Local::now() + Duration::days(90)).format("%Y-%m-%d").to_string()
}

fn bmi_category(bmi: f64) -> String {
    if bmi < 18.5 { "Underweight".to_string() }
    else if bmi < 25.0 { "Normal".to_string() }
    else if bmi < 30.0 { "Overweight".to_string() }
    else { "Obese".to_string() }
}

// ─── Main ────────────────────────────────────────────────────────────────────

fn main() {
    let mut app = WeightTracker::new();
    app.run();
}
